### PR TITLE
167190672 editing own email

### PR DESCRIPTION
**Pull request by jkorpelin**
# Added
* Possibility for users to edit their own email.

### CODE REVIEW

File: ote/src/cljs/ote/views/user.cljs [hunk 1]
```diff
@@ -25,76 +26,80 @@
...
+            {:name :email :type :string :autocomplete "email" :required? true
+             :full-width? true :placeholder (tr [:register :placeholder :email])
+             :validate [(fn [data _]
+                          (when (not (user/email-valid? data))
+                            (tr [:common-texts :required-field])))
+                        (fn [data _]
+                          (when (and email-taken (email-taken data))
+                            (tr [:register :errors :email-taken])))]
+             :should-update-check form/always-update}
+            (form/info "Huom! Jos muokkaat sähköpostiosoitetta tulee sinun vahvistaa uusi sähköpostiosoite ennen kuin voit kirjautua uudelleen sisään.")
```
Comment (rckopteri) [inline ote/src/cljs/ote/views/user.cljs]: Vois lisätä käännöstagin taakse

File: ote/src/cljs/ote/views/user_edit.cljs [hunk 2]
```diff
@@ -0,0 +1,84 @@
+(ns ote.views.user-edit
```
Comment (rckopteri) [inline ote/src/cljs/ote/views/user_edit.cljs]: Should this be only for admins? Users have their own page for editing their data

File: ote/src/cljs/ote/views/user_edit.cljs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+               :full-width? true :placeholder (tr [:register :placeholder :email])
+               :validate [(fn [data _]
+                            (when (not (user/email-valid? data))
+                              (tr [:common-texts :required-field])))
+                          (fn [data _]
+                            (when (and email-taken (email-taken data))
+                              (tr [:register :errors :email-taken])))]
+               :should-update-check form/always-update}
+              (if email-confirmed?
+                (form/info "Käyttäjä on varmentanut sähköpostiosoitteen.")
```
Comment (rckopteri) [inline ote/src/cljs/ote/views/user_edit.cljs]: localization

File: ote/src/cljs/ote/views/user_edit.cljs (hunk 2 above)
Comment (jkorpelin) [reply ote/src/cljs/ote/views/user_edit.cljs]: Yes only for admins

File: ote/src/cljs/ote/views/user_edit.cljs (hunk 3 above)
Comment (jkorpelin) [reply ote/src/cljs/ote/views/user_edit.cljs]: Yes it's only for admins so no localizations needed.

File: ote/src/clj/ote/services/users.clj [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+      (let [{new-email :email
+             new-username :username} form-data
+            username-taken? (and (not= (:username user) new-username)
+                              (username-exists? db {:username new-username}))
+            language (or (:language form-data) :fi)
+            email-changed? (not= (:email user) new-email)
+            email-taken? (and email-changed?
+                           (email-exists? db {:email new-email}))
+            login-info (first (login/fetch-login-info db {:email (:email user)}))
+            password-incorrect? (if admin?
```
Comment (philosotractor) [inline ote/src/clj/ote/services/users.clj]: A matter of taste, & this was already discussed, in any case making a note:
negations and inversions should be avoided in naming for better clarity.
```suggestion
            password-correct? (if admin?
                                             ( ...code...)
                                             false)
```

**Pull request closed without merging**